deploy-agent/deployd/common/stats.py

```diff
@@ -28,69 +49,287 @@ def __exit__(self, stats, sample_rate=1.0, tags=None):
+        """ add default tags to stats
+            :param: tags as dict
+            return: dict
+        """
+        if not __version__:
+            # defensive case, should not be hit
+            return tags
+        if __version__ and not tags:
+            tags = dict()
+        tags['deploy_agent_version'] = __version__
+        return tags
+
+    @staticmethod
+    def _parse_stat(mtype=None, name=None, value=None, sample_rate=None, tags=None):
+        """ return Stat for given kwargs """
+        return Stat(mtype=mtype,
+                    name=name,
+                    value=value,
+                    sample_rate=sample_rate,
+                    tags=tags)
+
+    def _send_mtype(self):
+        """ send metric to sc using corrected
+            calls per metric type
+        """
+        func = getattr(sc, self.stat.mtype)
+        func_v2 = getattr(sc_v2, self.stat.mtype)
+
+        if self.stat.mtype == 'increment':
+            func(self.stat.name, self.stat.sample_rate, self.stat.tags)
+            func_v2(self.stat.name, self.stat.sample_rate, self.stat.tags)
+        elif self.stat.mtype == 'gauge' or self.stat.mtype == 'timing':
+            func(self.stat.name, self.stat.value, sample_rate=self.stat.sample_rate, tags=self.stat.tags)
+            func_v2(self.stat.name, self.stat.value, sample_rate=self.stat.sample_rate, tags=self.stat.tags)
+
+    def _send(self):
+        """ send metric to sc """
+        try:
+            self._send_mtype()
+        except Exception as error:
+            log.error('unable to send metric: {}'.format(error))
+
+    def _flush_cache(self):
+        """ read from cache, send every metric, truncate cache """
+        log.warning('flushing metrics from cache')
+        for stat in self.cache.read():
+            if not stat.deserialize():
+                # unable to parse stat, skip
+                log.error('unable to parse stat: {}'.format(stat))
+                continue
+            self.stat = stat
+            self._send()
+        # truncate cache file after flush
+        self.cache.truncate()
+
+    def send_context_timer(self, name, value, sample_rate=None, tags=None):
+        """ convert a context_timer to timing
+            for cacheability
+        """
+        self.send(mtype='timing', name=name, value=value, sample_rate=sample_rate, tags=tags)
+
+    def send(self, mtype=None, name=None, value=None, sample_rate=None, tags=None):
+        """ add default tags, send metric, write to, or flush cache
+            depending on health check """
+        tags = self._add_default_tags(tags)
+        if self.is_healthy():
+            # trigger flush cache first
+            if self.cache and (self.cache.exists() and not self.cache.is_empty()):
+                self._flush_cache()
+            # send metric
+            self.stat = self._parse_stat(mtype=mtype, name=name, value=value, sample_rate=sample_rate, tags=tags)
+            self._send()
```

Comment:
pass stat as parameter to seld._send(stat) instead of defining member variable.

Reply:
The existing logic also reads from this member variable in other places as well, so changing this to a param means a lot of passing it around where it's needed